src/System.Private.CoreLib/src/System/Diagnostics/Eventing/EventPipePayloadDecoder.cs

```diff
@@ -110,16 +110,29 @@ internal static object[] DecodePayload(ref EventSource.EventMetadata metadata, R
                 }
                 else if (parameterType == typeof(string))
                 {
-                    ReadOnlySpan<char> charPayload = MemoryMarshal.Cast<byte, char>(payload);
-                    int charCount = charPayload.IndexOf('\0');
-                    if (charCount < 0)
+                    // Try to find null terminator (0x00) from the byte span
+                    // NOTE: we do this by hand instead of using payload.IndexOf('\n') because payload may be unaligned due to 
+                    // mixture of different types being stored in the same buffer. (see eventpipe.cpp:CopyData)
+                    int byteCount = -1;
+                    for (int j = 1; j < payload.Length; j+=2)
                     {
+                        if (payload[j-1] == (byte)(0) && payload[j] == (byte)(0))
+                        {
+                            byteCount = j+1;
+                            break;
+                        }
+                    }
+
+                    ReadOnlySpan<char> charPayload;
+                    if (byteCount < 0)
+                    {
+                        charPayload = MemoryMarshal.Cast<byte, char>(payload);
                         payload = default;
                     }
                     else
                     {
-                        charPayload = charPayload.Slice(0, charCount);
-                        payload = payload.Slice((charCount + 1) * sizeof(char));
+                        charPayload = MemoryMarshal.Cast<byte, char>(payload.Slice(0, byteCount));
```

Comment:
```suggestion
                        //byteCount includes the null terminator, which we do not want in the managed string object
                        charPayload = MemoryMarshal.Cast<byte, char>(payload.Slice(0, byteCount - sizeof(char));
```

byteCount includes the null terminator bytes, but managed string buffers aren't canonically null-terminated. For an example, this code snippet will fail the assert evaluating 4 == 3

            char[] catChars = { 'c', 'a', 't', (char)0 };
            ReadOnlySpan<char> span = new ReadOnlySpan<char>(catChars);
            string cat = new string(span);
            string cat2 = "cat";
            Debug.Assert(cat.Length == cat2.Length);